js/src/common/extend.ts

```diff
@@ -19,23 +26,27 @@
  *   // something that needs to be run on creation and update
  * });
  *
- * @param {object} object The object that owns the method
- * @param {string|string[]} methods The name or names of the method(s) to extend
- * @param {function} callback A callback which mutates the method's output
+ * @param object The object that owns the method
+ * @param methods The name or names of the method(s) to extend
+ * @param callback A callback which mutates the method's output
  */
-export function extend(object, methods, callback) {
+export function extend<T extends object, K extends KeyOfType<T, Function>>(
+  object: T,
+  methods: K | K[],
+  callback: (val: ReturnType<T[K]>, ...args: Parameters<T[K]>) => void
+) {
   const allMethods = Array.isArray(methods) ? methods : [methods];
 
-  allMethods.forEach((method) => {
-    const original = object[method];
+  allMethods.forEach((method: K) => {
+    const original: Function | undefined = object[method];
```

Comment:
Can we not replace `Function | undefined` with `T[K]`? `Function` is like `any`, so it's best not to use it if we can.

Same for `override`.

Reply:
`.apply` doesn't exist on `T[K]`, for whatever reason

Reply:
> `.apply` doesn't exist on `T[K]`, for whatever reason

Aren't we enforcing that `T[K] extends Function` though? Not a blocker, just curious :)

Reply:
You'd think so, but `T[K] extends Function` actually ends up being true when `T[K]` is a function, **and** when it is a subtype of `Function`.

This is something I ran into when playing with deep cloning objects: https://github.com/microsoft/TypeScript/issues/45492